### PR TITLE
[4.x] Sorts AnimationNodeBlendTree node connections for deterministic output

### LINKED ISSUE #42788: Nondeterministic .tscn data from AnimationBlendTreeNode
**Godot version:**
3.2.3

**OS/device including version:**
Windows 10, 1904.508

**Issue description:**
The order of the triplets in the "node_connections" array is nondeterministic, making it annoying to version control files with animaton blend trees as they may change without any actual alterations to the scene.

In my specific case, the diff often looks like this:
image
The array contains the same triplets but in different order.

**Steps to reproduce:**
* Create a scene with a larger animation tree (mine has 9 nodes with 8 connections) and save it.
* Commit the file to version control (GIT).
* Restart Godot, open and save the same file.
* Check the version control diff.
* Retry a few times from step 3 if no diff occurs.

However, it was easy to find the likely cause of the bug in the C++ code:
https://github.com/godotengine/godot/blob/00949f0c5fcc6a4f8382a4a97d5591fd9ec380f8/scene/animation/animation_blend_tree.cpp#L1122-L1138
https://github.com/godotengine/godot/blob/00949f0c5fcc6a4f8382a4a97d5591fd9ec380f8/scene/animation/animation_blend_tree.cpp#L1028-L1041

The .tscn array is created by iterating the elements of a Map, which order is undefined (unless I'm mistaken about the specifics of Godot's Map class). At the very least, the order of the Map elements are somewhat dependent on the insertion order, which may in turn be nondeterministic. Relying on the element order of the map here seems in any case suboptimal for consistent .tscn data.

**Suggested fix:**
The element order needs to be sorted in some way. Alphabetical sorting by each NodeConnection field in turn seems as good as any.

```C++
// I'm not very familiar with Godot's own primitives, but something like this:
int comp_nc(const NodeConnection &a, const NodeConnection &b) const {
  return a.input_node.compare(b.input_node) < 0 || a.input_index < b.input_index || a.output_node.compare(b.output_node) < 0;
}
// ...
bool AnimationNodeBlendTree::_get(const StringName &p_name, Variant &r_ret) const {
  //...
  List<NodeConnection> nc;
  get_node_connections(&nc);
  std::sort(nc.begin(), nc.end(), &comp_nc);
  //...
}
```

### PR DESCRIPTION
**Pull request by LoparPanda**
This should resolve #42788.

Is there a way to pull off the sort without the temporary `String` copies? I wasn't sure how `StringName` is set up but the `==` and `<` operators seem to be just comparing the inner `_data` pointers, so unless I missed something, the pointers would have to be pre-sorted in memory for that to be used for comparison.

Mirroring this in a 3.x PR as well.

### CODE REVIEW

File: scene/animation/animation_blend_tree.cpp [hunk 1]
```diff
@@ -1328,6 +1328,24 @@ void AnimationNodeBlendTree::get_node_connections(List<NodeConnection> *r_connec
 			}
 		}
 	}
+	// Keep node connections sorted for deterministic scene file output.
+	struct Comparator {
+		bool operator()(const NodeConnection &a, const NodeConnection &b) const {
```
Comment (AThousandShips) [inline scene/animation/animation_blend_tree.cpp:1333]:
```suggestion
		bool operator()(const NodeConnection &p_a, const NodeConnection &p_b) const {
```
We prefix arguments with `p_`

File: scene/animation/animation_blend_tree.cpp [hunk 2]
```diff
@@ -1328,6 +1328,24 @@ void AnimationNodeBlendTree::get_node_connections(List<NodeConnection> *r_connec
...
+			String a_input_node(a.input_node);
+			String b_input_node(b.input_node);
+			if (a_input_node < b_input_node) {
+				return true;
+			} else if (a_input_node == b_input_node) {
```
Comment (AThousandShips) [inline scene/animation/animation_blend_tree.cpp:1338]: Better to compare the original `StringName`, much faster

Comment (AThousandShips) [reply scene/animation/animation_blend_tree.cpp:1338]: I'd also suggest using `StringName::AlphCompare`

Comment (LoparPanda) [reply scene/animation/animation_blend_tree.cpp]: awesome, that's more efficient than the temporary strings and seems to work fine as well, thanks! :) will update 3.x PR momentarily as well.

**Comment by fire** [comment]
Can you read the https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html? In particular the squashed guideline.

I'm interested in reviewing this. @TokageItLab too.

**Comment by LoparPanda** [comment]
> Can you read the https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html? In particular the squashed guideline.
> 
> I'm interested in reviewing this. @TokageItLab too.

Done! Should be all squashed now on both PRs.

**Review by AThousandShips — APPROVED** [review]
LGTM, haven't tested but makes sense

**Review by TokageItLab — CHANGES_REQUESTED** [review]
This problem is caused by the fact that the `HashMap nodes` do not have an order.

However, changing it to `RBMap nodes` did not solve the problem. When I look at the `RBMap nodes` when adding a node to `add_node()`, it appears that the insertion into the RBMap is not done uniquely by key. @reduz Do you know anything about this?

So why is this problem not happening in other resources?

For example, VisualShader also uses GraphNode, but VisualShader's connections are stored for each node in a child of the `"nodes/"` path.

On the other hand, the AnimationBlendTree stored it in an ordered array called `"node_connections"`, even though the nodes are unordered. However, since the nodes are sorted and stored in `.tres`, this problem can be solved by storeing `"node_connections"` as children of `"nodes"` means `"nodes/XXX/connections"`. (Then, conversion code is needed for getter and setter for compatibility)

So, this PR's solution does not seem so smart, it may be applicable to 3.x, but at least, what needs to be sorted is the order of the `nodes`, not the `input_node`s. Also, it should only be needed when recording to `.tres` so it needs to be in `_get()` (this fix should not be necessary for BlendTreeEditor, etc.).

**Comment by AThousandShips** [comment]
> So, this PR's solution does not seem so smart

This isn't really a called for phrasing and isn't very constructive

**Comment by TokageItLab** [comment]
Let me rephrase.

Sorting where it is not needed is not good for performance. Also, this fix does not address the fundamental problem and is just a cop-out, and the implementation is inconsistent when compared to other classes.

**Comment by TokageItLab** [comment]
I tried to store it in "nodes/XXX/connections". Indeed, it has unique order. However, the result seems to be that it has to be kept separately as "node_connections". This is because there is a restriction that "connections" must be loaded after all other nodes.

Also, when I investigated again why `RBMap` works well in `VisualShader`, I found that it uses `int id` as the key. So, apparently there is a problem with `RBMap<StringName, V>`, but `RBMap<String, V>` definitely works well.

I sent https://github.com/godotengine/godot/pull/79595 which can supersede this PR.

**Comment by TokageItLab** [comment]
Superseded by #79595.

**Comment by AThousandShips** [comment]
Thank you for your work, and hope to see other contributions from you going forward, even if this didn't pan out!
